### PR TITLE
test: add unit tests for ForkContext

### PR DESCRIPTION
**Pull request by Kuldeep2822k**
#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### AI acknowledgment

- [ ] I did _not_ use AI to generate this PR.
- [x] (If the above is not checked) I have reviewed the AI-generated content before submitting.

#### What is the purpose of this pull request? (put an "X" next to an item)

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[ ] Add autofix to a rule
[ ] Add a CLI option
[ ] Add something to the core
[x] Other, please explain:

Added unit tests for `ForkContext` class (`lib/linter/code-path-analysis/fork-context.js`), which previously had no dedicated test coverage.

#### What changes did you make? (Give an overview)

Added `tests/lib/linter/code-path-analysis/fork-context.js` with 16 tests covering the full public API of `ForkContext`:

- **Static factories:** `newRoot()`, `newEmpty()` (both `shouldForkLeavingPath` true and false)
- **Getters:** `head`, `empty`, `reachable` (reachable, unreachable, and empty states)
- **Segment creation:** `makeNext()`, `makeUnreachable()`, `makeDisconnected()`
- **Mutation methods:** `add()` (basic + `mergeExtraSegments` path), `replaceHead()`, `addAll()`, `clear()`

#### Is there anything you'd like reviewers to focus on?

- The `makeNext()` test requires calling `CodePathSegment.markUsed()` on previous segments before invoking `makeNext()`, so that `flattenUnusedSegments` returns them correctly. Would appreciate a check that this setup accurately reflects real-world usage.
- The `add()` merge test passes more segments than `context.count` to exercise the `mergeExtraSegments` helper.

### CODE REVIEW

File: tests/lib/linter/code-path-analysis/fork-context.js [hunk 1]
```diff
@@ -0,0 +1,199 @@
+/**
+ * @fileoverview Tests for ForkContext.
+ * @author kuldeep kumar
+ */
+
+"use strict";
+
+const assert = require("node:assert");
+const ForkContext = require("../../../../lib/linter/code-path-analysis/fork-context");
+const IdGenerator = require("../../../../lib/linter/code-path-analysis/id-generator");
```
Comment (DMartens) [inline tests/lib/linter/code-path-analysis/fork-context.js]: This unit test file should only test `ForkContext` and not depend on `IdGenerator` or `CodePathSegment`.
`IdGenerator` and `CodePathSegment` should be stubbed so this unit test file is uncoupled to changes to them.
For this you need to provide a minimal implementation (without edge cases) which has the same behaviour,
for example for the id generator:
```javascript
const idGenerator = {
	prefix: "",
	n: 0,
	next() {
		this.n++;
		return this.n;
	}
}
```

File: tests/lib/linter/code-path-analysis/fork-context.js [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+const CodePathSegment = require("../../../../lib/linter/code-path-analysis/code-path-segment");
+
+describe("ForkContext", () => {
```
Comment (DMartens) [inline tests/lib/linter/code-path-analysis/fork-context.js:78]: A test suite for the constructor is missing.

File: tests/lib/linter/code-path-analysis/fork-context.js [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+	let idGenerator;
+
+	beforeEach(() => {
+		idGenerator = new IdGenerator("s");
+	});
+
+	describe("newRoot()", () => {
+		it("should create a new root context with one initial segment", () => {
```
Comment (DMartens) [inline tests/lib/linter/code-path-analysis/fork-context.js]: This test case should be split up into two test case: one for `upper` and `count` testing the created context.
The other for the `segmentsList` created by `context.add`.

File: tests/lib/linter/code-path-analysis/fork-context.js [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+			const context = ForkContext.newEmpty(parentContext, true);
+
+			assert.strictEqual(context.upper, parentContext);
+			assert.strictEqual(context.count, 2);
+			assert.strictEqual(context.segmentsList.length, 0);
+			assert.strictEqual(context.empty, true);
+		});
+	});
+
+	describe("properties", () => {
```
Comment (DMartens) [inline tests/lib/linter/code-path-analysis/fork-context.js]: Each getter (`head`, `reachable` and `empty`) should have their own test suite.
Can you please also add test cases for `empty`.

File: tests/lib/linter/code-path-analysis/fork-context.js [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+			context.add([seg2]);
+
+			/*
+			 * segmentsList has 3 elements: [root], [seg1], [seg2]
+			 * We need to mark segments used so that flattenUnusedSegments returns them
+			 */
+			CodePathSegment.markUsed(seg1);
+			CodePathSegment.markUsed(seg2);
+
+			const nextSegments = context.makeNext(-2, -1);
```
Comment (DMartens) [inline tests/lib/linter/code-path-analysis/fork-context.js:265]: The normalization of negative values for the start and end index should be its own test case.
In this case this would be equal to `context.makeNext(1, 2);`.
The same for the other test cases of `make*`.

File: tests/lib/linter/code-path-analysis/fork-context.js [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+		it("should create disconnected segments", () => {
+			const context = ForkContext.newRoot(idGenerator);
+			const nextSegments = context.makeDisconnected(-1, -1);
+			assert.strictEqual(nextSegments.length, 1);
+			assert.strictEqual(nextSegments[0].allPrevSegments.length, 0); // Disconnected doesn't connect
+			assert.strictEqual(nextSegments[0].reachable, true); // Inherits reachable from prev segments
+		});
+	});
+
+	describe("add()", () => {
```
Comment (DMartens) [inline tests/lib/linter/code-path-analysis/fork-context.js:351]: A test case for the assertion is missing.

File: tests/lib/linter/code-path-analysis/fork-context.js [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+
+			assert.strictEqual(context.segmentsList.length, 2);
+			assert.strictEqual(context.head.length, 1);
+			assert.strictEqual(context.head[0].allPrevSegments.length, 2);
+			assert.ok(context.head[0].allPrevSegments.includes(seg1));
+			assert.ok(context.head[0].allPrevSegments.includes(seg2));
+		});
+	});
+
+	describe("replaceHead()", () => {
```
Comment (DMartens) [inline tests/lib/linter/code-path-analysis/fork-context.js:396]: Two test cases are missing
- one for the assertion
- one that only the head is modified and not over elements

File: tests/lib/linter/code-path-analysis/fork-context.js [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+			const context = ForkContext.newRoot(idGenerator);
+			const newSegment = CodePathSegment.newNext("s2", [context.head[0]]);
+			context.replaceHead([newSegment]);
+
+			assert.strictEqual(context.segmentsList.length, 1);
+			assert.strictEqual(context.head[0], newSegment);
+		});
+	});
+
+	describe("addAll()", () => {
```
Comment (DMartens) [inline tests/lib/linter/code-path-analysis/fork-context.js:458]: A test case for the assertion is missing.

**Comment by Kuldeep2822k** [comment]
I've addressed all of your feedback in the latest commit:

- **Decoupled tests:** Removed `IdGenerator` and `CodePathSegment` dependencies, replacing them with minimal stubs so the test is isolated.
- **Constructor Tests:** Added a dedicated `describe("constructor")` test suite.
- **`newRoot()` Test:** Split the test case into two distinct assertions as requested.
- **Getters:** Separated `head`, `reachable`, and `empty` into their own `describe` blocks and added the missing `empty` test cases.
- **Negative Index Normalization:** Created isolated test cases verifying negative indices for `makeNext`, `makeUnreachable`, and `makeDisconnected`.
- **Missing Assertions:** Added assertion tests for `add()`, `replaceHead()`, and `addAll()`.
- **`replaceHead()` modification:** Added a test verifying that only the head element is modified.

Let me know if anything else needs adjusting!

File: tests/lib/linter/code-path-analysis/fork-context.js [hunk 9]
```diff
@@ -162,24 +394,84 @@ describe("ForkContext", () => {
...
-			const newSegment = CodePathSegment.newNext("s2", [context.head[0]]);
+			const newSegment = createSegment("s2", [context.head[0]], true);
+
 			context.replaceHead([newSegment]);
 
 			assert.strictEqual(context.segmentsList.length, 1);
 			assert.strictEqual(context.head[0], newSegment);
 		});
+
+		it("should only modify the last element of segmentsList", () => {
```
Comment (DMartens) [inline tests/lib/linter/code-path-analysis/fork-context.js]: This describes an implementation detail that the `head` is the last element of the segmentsList.
So this rather should be "should only modify the head".

File: tests/lib/linter/code-path-analysis/fork-context.js [hunk 10]
```diff
@@ -7,22 +7,126 @@
...
+	};
+}
+
+/**
+ * Marks a stub segment as used, mirroring the minimal behaviour needed
+ * for ForkContext's internal flattenUnusedSegments to work correctly.
+ * @param {Object} segment The segment to mark as used.
+ * @returns {void}
+ */
+function markUsed(segment) {
```
Comment (DMartens) [inline tests/lib/linter/code-path-analysis/fork-context.js]: This function is unneeded as it can be replaced with `segment.internal.used = true` for the current test cases.

**Review by DMartens — APPROVED** [review]
Changes LGTM, thanks.
